### ISSUE TITLE
List of all Pexec() problems

### ISSUE THREAD

**Issue by mikrosk** [atarijookie/ce-atari#162]
I'd like to take a look at the original Pexec() implementation (i.e. reverting the disk image hack by Jookie) sometime in near future. Can anyone remember problems with the original implementation? In other words, the problems which led Jookie to implement his workaround?

(I don't mean claims like "it didn't work in TOS 1.00", more like what exactly)

The original implementation was badly designed, I had expected it to be quick & simple handler and it has become a major pain. That doesn't mean it can't be reimplemented / redesigned.

Of course, this would need testing and volunteers so don't worry, it wont get into production code any time soon. I just want to gather as much information as possible.

**Comment by atarijookie** [comment]
- TSR (drivers) had issues, as Ptermres() didn't work correctly - either didn't tell TOS correctly what memory to not use anymore, or something similar
- nested Pexec() didn't work - I have only one set of pointer and states there, so one PRG calling another PRG and then another would definitelly cause some misbehavior

My guess is that TOS has some variables inside, which store some vital info which prevent us from nice impementation of Pexec(), but I might be wrong on this one (it seems that emulators do it just fine). Also it's not possible to debug these issues - either it works, or it doesn't and no-one knows why ;) Sure, you could trace it within emulator, but you can't connect the real CE device there...

**Comment by mikrosk** [comment]
Yeah, another (more mature) approach to debugging is one of my goals, too. Because let's face it, if I replace trap#1 vector with my own and redirect all calls through your code, there's no way I wont be able to debug this kind of scenarios (incl. different TOS versions etc). As I said, it was just implemented in rush, I didn't think of debug capabilities and similar stuff very much.

@ggnkua could you contribute with some of your painful experiences? ;)

**Comment by miniupnp** [comment]
I'm thinking, why not look into the EmuTOS implementation of Pexec() ?
